### PR TITLE
Move legality

### PR DESCRIPTION
**Pull request by gwenllian1**
- Only legal moves can now be made
- Created 'legality' structure for moves to indicate why they are illegal
- Updated tests that failed because the moves were illegal
- Wrote tests for move legality

Kat + Llio.

### CODE REVIEW

**Review by nocommentcode — CHANGES_REQUESTED** [review]
Great stuff guys, 
A few of my comments are irrelevant because the commiting is kind of all over the place so ignor the comments that you already addressed.
A few other things:
- In chess engine's generate pgn function, we are still calling chess.forceMove
- in the chess engine tests, for some reason the tests are still passing but they should be failing... Since we are not allowing illegal moves anymore, the illegal  move tests should be failing but they are not, i dont understand why

File: packages/TorneloScoresheet/src/hooks/appMode/recordingState.ts [hunk 1]
```diff
@@ -205,6 +205,21 @@ export const makeUseRecordingState =
...
+        if (!appModeState.pairing.positionOccurances) {
+          appModeState.pairing.positionOccurances = {};
+        }
+
+        appModeState.pairing.positionOccurances[key] =
+          key in appModeState.pairing.positionOccurances
+            ? (appModeState.pairing.positionOccurances[key] || 0) + 1
+            : 1;
+
+        console.log(appModeState.pairing.positionOccurances);
```
[the remark this answers is not in the document]
Comment (nocommentcode) [reply packages/TorneloScoresheet/src/hooks/appMode/recordingState.ts]: i think this was fixed in a later commit

File: packages/TorneloScoresheet/src/hooks/appMode/recordingState.ts [hunk 2]
```diff
@@ -68,80 +69,95 @@ const getCurrentFen = (moveHistory: ChessPly[]): string => {
...
-};
-
 export const makeUseRecordingState =
   (
     context: React.Context<
       [AppModeState, React.Dispatch<React.SetStateAction<AppModeState>>]
     >,
   ): (() => recordingStateHookType | null) =>
   (): recordingStateHookType | null => {
+    const [, showError] = useError();
```
[the remark this answers is not in the document]
Comment (nocommentcode) [reply packages/TorneloScoresheet/src/hooks/appMode/recordingState.ts]: i think this was fixed in a later commit

File: packages/TorneloScoresheet/src/hooks/appMode/recordingState.ts [hunk 3]
```diff
@@ -142,6 +92,123 @@ export const makeUseRecordingState =
...
+      moveLegality.inCheck = chessEngine.inCheck(fen);
+      moveLegality.inDraw = chessEngine.inDraw(fen);
+      moveLegality.inCheckmate = chessEngine.inCheckmate(fen);
+      moveLegality.insufficientMaterial = chessEngine.insufficientMaterial(fen);
+      moveLegality.inStalemate = chessEngine.inStalemate(fen);
+      if (moveHistory[index]) {
+        moveHistory[index]!.legality = moveLegality;
+        return moveHistory;
+      }
+      return;
```
[the remark this answers is not in the document]
Comment (nocommentcode) [reply packages/TorneloScoresheet/src/hooks/appMode/recordingState.ts:121]: just to be clear, this function should just take in a fen and return a moveLegality object, and in processPlayerMove, when we instanciate the next ply, we will have moveLegality:checkMoveLegality(nexFen)

**Comment by nocommentcode** [comment]
Also if we are in 5 fold repetition, how is the arbiter mode supposed to detect this? Since we arent setting the 5 fold repetition legality check to true in all 5 plys in the code (unless i missed it)?

File: packages/TorneloScoresheet/src/pages/GraphicalRecording/PromotionSheet.tsx [hunk 4]
```diff
@@ -20,7 +20,6 @@ const PromotionSheet: React.FC<PromotionSheetProps> = ({
       onPress: () => handleSelectPromotion(PieceType.Queen),
     },
     { icon: ROOK, onPress: () => handleSelectPromotion(PieceType.Rook) },
-    { icon: PAWN, onPress: () => handleSelectPromotion(PieceType.Pawn) },
```
[the remark this answers is not in the document]
Comment (gwenllian1) [reply packages/TorneloScoresheet/src/pages/GraphicalRecording/PromotionSheet.tsx:23]: image
we googled it and the general vibe was that you had to promote a pawn

File: packages/TorneloScoresheet/src/hooks/appMode/recordingState.ts [hunk 5]
```diff
@@ -142,6 +92,123 @@ export const makeUseRecordingState =
...
+    /**
+     * Skips a player's turn
+     * Will Return the next move history array
+     * @param moveHistory ChessPly array of past moves
+     * @returns new moveHistory array
+     */
+    const skipPlayerTurn = (moveHistory: ChessPly[]): Result<ChessPly[]> => {
+      if (gameInFiveFoldRepetition()) {
+        return fail('Illegal Move');
+      }
```
[the remark this answers is not in the document]
Comment (gwenllian1) [reply packages/TorneloScoresheet/src/hooks/appMode/recordingState.ts:198]: If we don't include the check then players can add skips when the game is already in fivefold repetition which they shouldn't be allowed to do.

File: packages/TorneloScoresheet/src/pages/GraphicalRecording/PromotionSheet.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (nocommentcode) [reply packages/TorneloScoresheet/src/pages/GraphicalRecording/PromotionSheet.tsx:23]: oh true ignore me then

**Review by nocommentcode — CHANGES_REQUESTED** [review]
good stuff i found the problem with the tests, 
we need:
`expect(nextFenResult !== null).toBe(true);` on lines 1024 and 1055, this is required to make sure the engine is not returning null on a possible move. This will break the tests so you can fix the test cases that should not be passing

**Review by nocommentcode — APPROVED** [review]
Perfect gg!

**Review by ksov0001 — APPROVED** [review]
Approving again after fixing package lock for CI

**Review by ksov0001 — APPROVED** [review]
Triggering CI again

**Review by nocommentcode — APPROVED** [review]
fixing ci issue

**Comment by nocommentcode** [comment]
Closed as I am refractoring some sections of this pr

**Pull request closed without merging**